File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        int max_len = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) nums[i] = -1;
        }
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (map.containsKey(sum))
                max_len = Math.max(max_len, i - map.get(sum));
            else map.put(sum, i);
        }
        return max_len;
    }
}

File: 0525-contiguous-array/0525-contiguous-array.scala
object Solution {
   def findMaxLength(nums: Array[Int]): Int = {
    def go(sum: Int,  maxx: Int, map: Map[Int, Int], i: Int): Int = {
      if (i == nums.length) maxx
      else {
        val x = sum + (if (nums(i) == 1) 1 else -1)
        if (map.contains(x)) {
          val idx = map.getOrElse(x, 0)
          val len = i - idx
          go(x, math.max(maxx, len), map, i + 1)
        }
        else go(x, maxx, map + (x -> i), i + 1)
      }
    }
    go(0, 0, Map(0 -> -1), 0)
  }
}